[SelectPanel] Filter input border and loading/error panel min-height

jamieshark opened the pull request:
_Authors: Please fill out this form carefully and completely._

_Reviewers: By approving this Pull Request you are approving the code change, as well as its deployment and mitigation plans._
_Please read this description carefully. If you feel there is anything unclear or missing, please ask for updates._

### What are you trying to accomplish?
<!-- Provide a description of the changes. -->
This PR addresses a few design items towards the SelectPanel component (https://github.com/github/primer/issues/3875). Namely, it adds a divider below the filter input (if present), it moves the filter banner error to be above the input, and adds a min-height that corresponds to the overlay size for the loading / error states.

### Screenshots
<!-- Provide before/after screenshots, videos, or graphs for any visual changes; otherwise, remove this section -->

<img width="579" alt="Screenshot of loading state with min height" src="https://github.com/user-attachments/assets/2a3bccb1-92a5-42b8-8d20-9df5221db3c6">
<img width="634" alt="Screenshot of error state with min height" src="https://github.com/user-attachments/assets/d77e93d0-a280-4f31-bb61-d3ecbc711b6d">
<img width="598" alt="Screenshot of error banner" src="https://github.com/user-attachments/assets/29b96f7f-ba56-4311-b413-72ad49fef8be">
<img width="703" alt="Screenshot of no results" src="https://github.com/user-attachments/assets/357bb6ee-5fac-4677-b8f7-314469abf973">

### Integration
<!-- Does this change require any updates to code in production? -->

#### List the issues that this change affects.
<!--Every code change must address _at least 1_ issue. Fixes a bug, completes a task, every change
      should have a corresponding issue listed here. If one does not already exist, create one. -->

- Closes https://github.com/github/primer/issues/3879
- Closes https://github.com/github/primer/issues/3881
- Closes https://github.com/github/primer/issues/3880
- Closes https://github.com/github/primer/issues/3934#issue-2514579172

#### Risk Assessment
  <!-- Please select from one of the following and detail why this level was chosen -->

- [x] **Low risk** the change is small, highly observable, and easily rolled back.
- [ ] **Medium risk** changes that are isolated, reduced in scope or could impact few users. The change will not impact library availability.
- [ ] **High risk** changes are those that could impact customers and SLOs, low or no test coverage, low observability, or slow to rollback.

### What approach did you choose and why?
<!-- This section is a place for you to describe your thought process in making these changes.
     List any tradeoffs you made to take on or pay down tech debt.
     Identify any work you did to mitigate risk.
     Describe any alternative approaches you considered and why you discarded them. -->

### Anything you want to highlight for special attention from reviewers?
<!-- This is your chance to identify remaining risks and confess any uncertainties you may have about the correctness of the changes.
     Highlight anything on which you would like a second (or third) opinion.
     Keep in mind how many component uses cases may be affected by your changes when assessing risk. -->

### Accessibility
<!--
  You may remove this section and the "Accessibility" heading above _only_ if the changes in this pull request do not impact UI. Delete all those that don't apply.
  If there are any accessibility-related updates, please describe them here.
-->
- **Fixes axe scan violation** - This change fixes an existing [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation.
- **No new axe scan violation** - This change does not introduce any new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violations.
- **New axe violation** - This change introduces a new [axe scan](https://thehub.github.com/epd/engineering/dev-practicals/frontend/accessibility/readiness-routine/development/#axe-scans) violation. Please describe why the violation cannot be resolved below.

### Merge checklist

- [ ] Added/updated tests
- [ ] Added/updated documentation
- [ ] Added/updated previews (Lookbook)
- [x] Tested in Chrome
- [x] Tested in Firefox
- [x] Tested in Safari
- [ ] Tested in Edge

Take a look at the [What we look for in reviews](https://github.com/primer/react/blob/main/contributor-docs/CONTRIBUTING.md#what-we-look-for-in-reviews) section of the contributing guidelines for more information on how we review PRs.

jamieshark commented on app/components/primer/alpha/select_panel.rb:
```diff
@@ -388,7 +388,10 @@ def initialize(
         @system_arguments[:tag] = :"select-panel"
         @system_arguments[:preload] = true if @src.present? && preload?
-
+        @system_arguments[:classes] = class_names(
```
not sure if this is necessary since we're not actually using this class for styles, but thought it felt like good best practice, but happy to take it out if it is more misleading.

camertron commented on app/components/primer/alpha/select_panel.html.erb:
```diff
@@ -32,20 +41,12 @@
                   position: :absolute
                 }
               )) %>
+              <%= render(Primer::Alpha::ActionList::Divider.new(mb: 0)) %>
```
Hmm the `Divider` component isn't really designed to be used like this. Can we write a bit of CSS instead?

langermank replied:
Doesn't the header have a divider param? The `Dialog` has `show_divider`

camertron replied:
Oh dang it does! https://github.com/primer/view_components/blob/6348ede37a08d9b325aa24bd30156bd9a8377a08/app/components/primer/alpha/dialog/header.rb#L56

camertron commented on app/components/primer/alpha/select_panel.pcss:
```diff
@@ -0,0 +1,3 @@
+.SelectPanel-loadingPanel {
```
```suggestion
.SelectPanel-loadingSpinner {
```

langermank replied:
I think this will still look odd if there aren't many results, but I guess it's still an improvement? Maybe we could add a preview to show how that would look (so setting the size to `medium_portrait` for example, and then only return one or two results

jamieshark replied:
@camertron - I don't know if `loadingSpinner` is totally accurate because this style also applies to the error message when the spinner is hidden. Do you still want the class name change?

camertron replied:
Hmm... alright let's keep `loadingPanel` then 👍

camertron reviewed (changes requested):
Nice! I'll let @langermank give the final go-ahead tho 😄

jamieshark commented:
@langermank - I've added a `limit_items` param for the playground where you can preview the various sizes with a specified number of items. Here's an example of `:medium_portrait` size with `2` items showing after load.

http://primer-view-components-preview-3044.eastus.azurecontainer.io/view-components/lookbook/inspect/primer/alpha/select_panel/playground?size=medium_portrait&limit_items=2

jamieshark commented:
@maximedegreve - the error message shown in the playground is only the default shown by the element. The consumer can provide custom `error_content` if they like, so I don't think a separate slot is needed, since the consumer can just provide that directly as the `error_content` slot. I would recommend creating another issue for the loading-state slot, so we are not preventing the other 3 issues from being closed as part of this work.

With regards to the other changes, I've added a little padding between the icon and text (the spec specifies `14px` but in the interest of keeping with the system I've given `mt-3` which compiles to `16px) and updated the font size of the filter banner.

maximedegreve commented:
Noticed a few small things. Some might be able to be addressed in this issue others might need opening of new issues.

1. The margin above the title should only be 8px and not 16px.

<img width="705" alt="" src="https://github.com/user-attachments/assets/9df6a529-3eb1-4d3a-9cf9-5057fd65d77e">

2. The example below closes the panel when I initiate a filter query and press enter. Since the focus is the input and not on an item this is unexpected and incorrect. The cursor is also in the wrong spot after opening it again. I can't remember 100% sure anymore but I think it should wipe the filter value when closing the panel.

https://github.com/user-attachments/assets/35aa8dd1-4ff3-4b7f-9027-ab4c5599da3a

3. Multi select should have visual checkboxes (not semantically) to visually indicate to users it won't close the panel when making a multiple selection. Single select is correct to use the checkmark but multi select isn't. See the [screenshots in our docs](https://primer.style/components/selectpanel).

4. The empty state is also missing a min-height similar to the error/loading states

<img width="849" alt="Screenshot 2024-09-09 at 10 29 25" src="https://github.com/user-attachments/assets/dd6d462c-8a6d-4264-a251-6f16d15df827">

5. The playground example changes in width after results are loaded in. This causes a layout swift we want to avoid.

https://github.com/user-attachments/assets/64c0c83a-0f77-4481-9bb9-b225e13f651a

jamieshark commented:
@maximedegreve 

1. I have addressed this with my recent changes, though I would like to note that the spec specified `14px` and I'm wondering what can be done to mitigate this confusion in the future.
2. I have created another issue for this: https://github.com/github/primer/issues/3932
3. I have created another issue for this: https://github.com/github/primer/issues/3933
4. I have addressed this in my changes.
5. I believe this is happening because the overlay size is set to `:auto`. If you would like to avoid width changes, it might help to select a different size. Otherwise, there's not really any way for us to know the exact width of the content that's coming in.

maximedegreve reviewed (approved):
Approving since the other reported issues will be saved as separate issues.

kendallgassner commented on demo/app/controllers/select_panel_items_controller.rb:
```diff
@@ -35,9 +35,10 @@ def index
+                res = SELECT_PANEL_ITEMS.select do |item|
                   [item[:title], item[:description]].join(" ").downcase.include?(query)
                 end
+                params.fetch(:limit_items, -1).to_i >= 0 ? res.take(params[:limit_items].to_i) : res
```
Can we throw this in a reivew labs -- I want to make sure there are no issues with single select, limits, and searching

kendallgassner commented:
✅ For single-select, I tested that if a page limit is added the selected state is updated correctly as Items are loaded/unloaded. 


![A video showing that the selected item state is retained as I filter using the SelectPanel's input.](https://github.com/user-attachments/assets/7517a822-713a-47e7-a957-5f087fba4756)

kendallgassner commented:
🔴 I did find one niche bug. If a selectedItem does not initial show due to a page limit... It is never selected. See video:



 
![A video where I change the page limit to 1. The selected item is not loaded. I then filter by the selectItems name showing it is not selected.](https://github.com/user-attachments/assets/c5010636-7a03-4be6-852d-0f5370913a0e)

kendallgassner reviewed (approved):
💟 ✅

camertron reviewed (approved):
Nice!
